packages/dd-trace/src/debugger/devtools_client/snapshot/collector.js

```diff
@@ -0,0 +1,153 @@
+'use strict'
+
+const session = require('../session')
+
+const LEAF_SUBTYPES = new Set(['date', 'regexp'])
+const ITERABLE_SUBTYPES = new Set(['map', 'set', 'weakmap', 'weakset'])
+
+module.exports = {
+  getRuntimeObject: getObject
+}
+
+// TODO: Can we speed up thread pause time by calling mutiple Runtime.getProperties in parallel when possible?
+// The most simple solution would be to swich from an async/await approach to a callback based approach, in which case
+// each lookup will just finish in its own time and traverse the child nodes when the event loop allows it.
+// Alternatively, use `Promise.all` or something like that, but the code would probably be more complex.
+
+async function getObject (objectId, maxDepth, depth = 0) {
+  const { result, privateProperties } = await session.post('Runtime.getProperties', {
+    objectId,
+    ownProperties: true // exclude inherited properties
+  })
+
+  if (privateProperties) result.push(...privateProperties)
+
+  return traverseGetPropertiesResult(result, maxDepth, depth)
+}
+
+async function traverseGetPropertiesResult (props, maxDepth, depth) {
+  // TODO: Decide if we should filter out non-enumerable properties or not:
+  // props = props.filter((e) => e.enumerable)
+
+  if (depth >= maxDepth) return props
+
+  for (const prop of props) {
+    if (prop.value === undefined) continue
+    const { value: { type, objectId, subtype } } = prop
+    if (type === 'object') {
+      if (objectId === undefined) continue // if `subtype` is "null"
+      if (LEAF_SUBTYPES.has(subtype)) continue // don't waste time with these subtypes
+      prop.value.properties = await getObjectProperties(subtype, objectId, maxDepth, depth)
+    } else if (type === 'function') {
+      prop.value.properties = await getFunctionProperties(objectId, maxDepth, depth + 1)
+    }
+  }
+
+  return props
+}
+
+async function getObjectProperties (subtype, objectId, maxDepth, depth) {
+  if (ITERABLE_SUBTYPES.has(subtype)) {
+    return getIterable(objectId, maxDepth, depth)
+  } else if (subtype === 'promise') {
+    return getInternalProperties(objectId, maxDepth, depth)
+  } else if (subtype === 'proxy') {
+    return getProxy(objectId, maxDepth, depth)
+  } else if (subtype === 'arraybuffer') {
+    return getArrayBuffer(objectId, maxDepth, depth)
+  } else {
+    return getObject(objectId, maxDepth, depth + 1)
+  }
+}
+
+// TODO: The following extra information from `internalProperties` might be relevant to include for functions:
+// - Bound function: `[[TargetFunction]]`, `[[BoundThis]]` and `[[BoundArgs]]`
```

Comment:
So you think we should include this info in the captured snapshot?

Reply:
I would love to sit down and see examples of each of those.

I think capture the BoundThis would be beneficial for debugging - as it would let the user know what the instance that was bounded. 

What I'm not sure if those also expose as locals and args (specifically for BoundArgs) - if they are duplicated, not sure this worth the effort. but if they are not captured, I think those values would improve the debug experience.